### PR TITLE
Typescript support

### PR DESCRIPTION
**Pull request by joseluisq**
## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

* [x] My code follows the code style of this project.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have read the **CONTRIBUTING** document.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.

### CODE REVIEW

**Comment by frenzzy** [comment]
@joseluisq thanks for the pull request! 🍻

Unfortunately this project does not use TypeScript and was written in pure JavaScript for simplicity. As a single maintainer I don't feel like I can commit to maintaining these typings over time. I appreciate the work you put into creating them to begin with, but stale typings seem inevitable as I make changes to the codebase given my lack of familiarity with TypeScript.

Please resubmit typings to [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) repo. I'd be happy to add a mention in the readme linking to your submission for other developers who want to find them.

**Comment by joseluisq** [comment]
> Unfortunately this project does not use TypeScript and was written in pure JavaScript for simplicity. As a single maintainer I don't feel like I can commit to maintaining these typings over time. I appreciate the work you put into creating them to begin with, but stale typings seem inevitable as I make changes to the codebase given my lack of familiarity with TypeScript.

Sure, I understand your reasons.

> Please resubmit typings to DefinitelyTyped repo. I'd be happy to add a mention in the readme linking to your submission for other developers who want to find them.

Yeah, I will try to redirect it to DefinitelyTyped because in my company we use Typescript and have chosen Hyperapp (for several reasons) and typings support. So I thought it's not a bad idea to add typings for `render` too.
Anyway, Thanks for the clarification.

**Comment by jorgebucaran** [comment]
@joseluisq I completely understand @frenzzy point of view! The good news is that a future version after V2 may include built-in support for renderToString, then I'd be happy to include your typings in the main repo.

Until then! 🙇

**Comment by joseluisq** [comment]
> @joseluisq I completely understand @frenzzy point of view! The good news is that a future version after V2 may include built-in support for renderToString, then I'd be happy to include your typings in the main repo.
> 
> Until then! 🙇

Sounds great 👍 

At the moment, I have prefered to add above `render` typings into my project directly.
Since Hyperapp is not into _DefinitelyTyped_ so it is not possible to reference
`/// <reference types="hyperapp" />` from other typings file like `render.d.ts`.
Because (in this case) [`withRender(nextApp)`](https://github.com/hyperapp/render/blob/master/src/node.js#L21) function depends on Hyperapp typings.

Complementing what it already explained on [this thread](https://github.com/hyperapp/hyperapp/issues/518#issuecomment-354920924), it would be great if Hyperapp could maintain typings more consistent across the @hyperapp organization in coming releases.

**Comment by jorgebucaran** [comment]
@joseluisq This is difficult to do because few of us actually use TS, but I'm going to try more! 💪

**Comment by joseluisq** [comment]
> @joseluisq This is difficult to do because few of us actually use TS, but I'm going to try more! 💪

Sure, I could contribute if you need a extra hand ⚡️

**Comment by frenzzy** [comment]
Btw, I don't mind to add typings to this repo if you are going to maintain them in the future :)

**Comment by joseluisq** [comment]
> Btw, I don't mind to add typings to this repo if you are going to maintain them in the future :)

Happy to hear about it! Let me know how can I help 💯

File: render.d.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+export as namespace render
+
+import { ActionsType, View } from 'hyperapp'
```
Comment (frenzzy) [inline render.d.ts]: Does this requires `hyperapp` to be in dependencies or peer/dev deps are enough?

File: test/ts/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{
+  "compilerOptions": {
+    "strict": true,
+    "noEmit": true,
+    "jsx": "react",
+    "jsxFactory": "h",
+    "baseUrl": "../..",
+    "typeRoots": [
+      "node_modules/@types/node/index.d.ts"
```
Comment (frenzzy) [inline test/ts/tsconfig.json]: Is this also requires `@types/node` to be in deps? Tests are not passing...

Comment (joseluisq) [reply test/ts/tsconfig.json]: Let me to review the tests again...

File: package.json [hunk 3]
```diff
@@ -51,6 +50,10 @@
     "rollup-plugin-uglify": "^3.0.0",
     "typescript": "^2.9.2"
   },
+  "dependencies": {
+    "@types/node": "^10.5.2",
+    "hyperapp": "^1.2.6"
```
Comment (joseluisq) [inline package.json]: Basically I added `@types/node` and `hyperapp` deps.

**Comment by frenzzy** [comment]
@jorgebucaran or maybe it is better to publish `hyperapp` typescript typings to DefinitelyTyped repo? To be able to publish typings of any hyperapp project, like this one, to DefinitelyTyped too.

Otherwise it requires to add all typings dependencies to the project and users have to install them even if they do not use typescript =/

All popular projects have typings in DefinitelyTyped by similar reasons, like `@types/express` `@types/node` and many more. The only exception is the libraries written in TypeScript itself. Thoughts?

**Comment by jorgebucaran** [comment]
I'd prefer to provide working types out of the box, like we do in Hyperapp. 

I think it's okay to take in @joseluisq changes once you are satisfied and if we run into issues in the future, e.g., when you want to change something in @hyperapp/render, you should just not forget to mention José so he can have a look.

I used to be more hesitant about TS in the past, but I think it's worth the trouble. If we can avoid forcing TS users to install another dep to add types, is a win for us.

**Comment by joseluisq** [comment]
> Otherwise it requires to add all typings dependencies to the project and users have to install them even if they do not use typescript =/

yep, so if you want to work with Typescript can you install typings on demand.

**Comment by joseluisq** [comment]
> I used to be more hesitant about TS in the past, but I think it's worth the trouble. If we can avoid forcing TS users to install another dep to add types, is a win for us.

@jorgebucaran right, I guess that v2 can address this on int the future.

**Comment by joseluisq** [comment]
In addition, Fee free to notify me or add me like reviewer.

**Comment by frenzzy** [comment]
> so if you want to work with Typescript you can install typings on demand.

Yes, but how to install dependencies with typings "on demand" but not always? Will moving them to `peerDependencies` section in package.json help?

Also would be cool to support Node.js since v6. So, can we use `>=6` or similar as a version number for `@types/node`?

**Comment by frenzzy** [comment]
@joseluisq great job, thanks!

> But consider to add some "info message" about peer deps if the users want to use TS.

FYI, at the moment npm throws a warning when a peer dependency is omitted:
```sh
npm WARN @hyperapp/render@2.0.0 requires a peer of @types/node@>=6.x but none is installed. You must install peer dependencies yourself.
```
Personally I think this is not cool because it sounds like a mandatory, but it is more like npm related issue: https://github.com/npm/npm/issues/20646

**Comment by joseluisq** [comment]
> Personally I think this is not cool because it sounds like a mandatory, but it is more like npm related issue: npm/npm#20646

Exactly, looks not good for `render` case. But I hope that in the future you consider to publish the Hyperapp typings to DefinitelyTyped (`@types/node` reason).

So, you keep out TS Devs of your codebase. (joke 😅)

File: package.json [hunk 4]
```diff
@@ -40,18 +42,23 @@
...
+    "typescript": "^2.9.2"
+  },
+  "peerDependencies": {
+    "@types/node": ">=6.x",
+    "hyperapp": "^1.2.6"
   },
   "scripts": {
     "lint": "eslint benchmark src test tools",
-    "test": "jest",
+    "test": "jest --coverage --no-cache && tsc -p test/ts",
```
Comment (frenzzy) [inline package.json]: Why do we need `--no-cache` here?

Comment (joseluisq) [reply package.json]: I'm not sure. https://github.com/facebook/jest/issues/6610

**Comment by joseluisq** [comment]
Cheers! 🥇 
Waiting for the next release!

**Comment by joseluisq** [comment]
Sure, reviewing...

File: render.d.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
+export as namespace render
```
Comment (frenzzy) [inline render.d.ts]: What the convention do we have `render` or `hyperappRender`? See https://github.com/hyperapp/logger/issues/18
Looks like we agreed to use `window.hyperappRender` should this namespace be the same?

Comment (joseluisq) [reply render.d.ts]: I sent a PR about this addressed at #11

**Comment by frenzzy** [comment]
Released as [v2.1.0](https://github.com/hyperapp/render/releases/tag/v2.1.0) :tada:

**Pull request merged**